I rebased on master but I think I'm now getting an unrelated failure bout `libindex`:
```
black --version
black, version 19.10b0
Checking black formatting
All done! ✨ 🍰 ✨
941 files would be left unchanged.
Checking black formatting DONE
flake8 --version
3.7.9 (flake8-comprehensions: 3.2.2, mccabe: 0.6.1, pycodestyle: 2.5.0, pyflakes: 2.1.1) CPython 3.7.6 on Linux
Linting .py code
##[error]./pandas/core/series.py:1025:9:F821:undefined name 'libindex'
##[error]./pandas/core/series.py:1108:17:F821:undefined name 'libindex'
Linting .py code DONE
```

--- a/doc/source/whatsnew/v1.1.0.rst
+++ b/doc/source/whatsnew/v1.1.0.rst
@@ -223,6 +223,8 @@ Other
 - Appending a dictionary to a :class:`DataFrame` without passing ``ignore_index=True`` will raise ``TypeError: Can only append a dict if ignore_index=True``
   instead of ``TypeError: Can only append a Series if ignore_index=True or if the Series has a name`` (:issue:`30871`)
 - Set operations on an object-dtype :class:`Index` now always return object-dtype results (:issue:`31401`)
+- Bug in :meth:`AbstractHolidayCalendar.holidays` when no rules were defined (:issue:`31415`)
+-
 
 .. ---------------------------------------------------------------------------
 

--- a/pandas/tests/tseries/holiday/test_calendar.py
+++ b/pandas/tests/tseries/holiday/test_calendar.py
@@ -98,3 +98,14 @@ def test_calendar_2031():
     Sat_before_Labor_Day_2031 = to_datetime("2031-08-30")
     next_working_day = Sat_before_Labor_Day_2031 + 0 * workDay
     assert next_working_day == to_datetime("2031-09-02")
+
+
+def test_no_holidays_calendar():
+    # Test for issue #31415
+
+    class NoHolidaysCalendar(AbstractHolidayCalendar):
+        pass
+
+    cal = NoHolidaysCalendar()
+    holidays = cal.holidays(Timestamp("01-Jan-2020"), Timestamp("01-Jan-2021"))
+    assert holidays.empty

--- a/pandas/tests/tseries/holiday/test_calendar.py
+++ b/pandas/tests/tseries/holiday/test_calendar.py
@@ -108,5 +108,5 @@ def test_no_holidays_calendar():
 
     cal = NoHolidaysCalendar()
     holidays = cal.holidays(Timestamp("01-Jan-2020"), Timestamp("01-Jan-2021"))
-    empty_series = Series(index=DatetimeIndex([]), dtype=object)
-    tm.assert_series_equal(holidays, empty_series)
+    empty_index = DatetimeIndex([])  # Type is DatetimeIndex since return_name=False
+    tm.assert_series_equal(holidays, empty_index)

--- a/pandas/tests/tseries/holiday/test_calendar.py
+++ b/pandas/tests/tseries/holiday/test_calendar.py
@@ -2,7 +2,7 @@ from datetime import datetime
 
 import pytest
 
-from pandas import DatetimeIndex, offsets, to_datetime
+from pandas import DatetimeIndex, Series, offsets, to_datetime
 import pandas._testing as tm
 
 from pandas.tseries.holiday import (
@@ -108,4 +108,5 @@ def test_no_holidays_calendar():
 
     cal = NoHolidaysCalendar()
     holidays = cal.holidays(Timestamp("01-Jan-2020"), Timestamp("01-Jan-2021"))
-    assert holidays.empty
+    empty_series = Series(index=DatetimeIndex([]), dtype=object)
+    tm.assert_series_equal(holidays, empty_series)

--- a/pandas/tseries/holiday.py
+++ b/pandas/tseries/holiday.py
@@ -7,7 +7,7 @@ import numpy as np
 
 from pandas.errors import PerformanceWarning
 
-from pandas import DateOffset, Series, Timestamp, date_range
+from pandas import DateOffset, DatetimeIndex, Series, Timestamp, concat, date_range
 
 from pandas.tseries.offsets import Day, Easter
 
@@ -406,17 +406,14 @@ class AbstractHolidayCalendar(metaclass=HolidayCalendarMetaClass):
         start = Timestamp(start)
         end = Timestamp(end)
 
-        holidays = None
         # If we don't have a cache or the dates are outside the prior cache, we
         # get them again
         if self._cache is None or start < self._cache[0] or end > self._cache[1]:
-            for rule in self.rules:
-                rule_holidays = rule.dates(start, end, return_name=True)
-
-                if holidays is None:
-                    holidays = rule_holidays
-                else:
-                    holidays = holidays.append(rule_holidays)
+            holidays = [rule.dates(start, end, return_name=True) for rule in self.rules]
+            if holidays:
+                holidays = concat(holidays)
+            else:
+                holidays = Series(index=DatetimeIndex([]), dtype=object)
 
             self._cache = (start, end, holidays.sort_index())
 

